src/Cardano/CoinSelection/Random.hs

```diff
@@ -49,66 +49,168 @@ import Data.Functor
+--      no further.
+--
+--      A selection with value /v1/ is considered to be an /improvement/ over a
+--      selection with value /v0/ if __all__ of the following conditions are
+--      satisfied:
+--
+--       * __Condition 1__: we have moved closer to the /ideal/ value:
+--
+--             abs (/ideal/ − /v1/) < abs (/ideal/ − /v0/)
+--
+--       * __Condition 2__: we have not exceeded the /maximum/ value:
+--
+--             /v1/ ≤ /maximum/
+--
+--       * __Condition 3__: when counting cumulatively across all outputs
+--       considered so far, we have not selected more than the /maximum/ number
+--       of UTxO entries specified by 'maximumInputCount'.
+--
+--  3.  __Creates a /change value/__ for the output, equal to the total value
+--      of the /final UTxO selection/ for that output minus the value /v/ of
+--      that output.
+--
+--  4.  __Updates the /accumulated coin selection/__:
+--
+--       * Adds the /output/ to 'outputs'.
+--       * Adds the /improved UTxO selection/ to 'inputs'.
+--       * Adds the /change value/ to 'change'.
+--
+-- This phase ends when every output has been processed, __or__ when the
+-- /remaining UTxO set/ has been exhausted, whichever occurs sooner.
+--
+-- = Termination
+--
+-- When both phases are complete, the algorithm terminates.
+--
+-- The /accumulated coin selection/ and /remaining UTxO set/ are returned to
+-- the caller.
+--
+-- === Failure Modes
+--
+-- The algorithm terminates with an __error__ if:
+--
+--  1.  The /total value/ of the initial UTxO set (the amount of money
+--      /available/) is /less than/ the total value of the output list (the
+--      amount of money /required/).
+--
+--      See: __'ErrUtxoBalanceInsufficient'__.
+--
+--  2.  The /number/ of entries in the initial UTxO set is /smaller than/ the
+--      number of requested outputs.
+--
+--      Due to the nature of the algorithm, /at least one/ UTxO entry is
+--      required /for each/ output.
+--
+--      See: __'ErrUtxoNotFragmentedEnough'__.
+--
+--  3.  Due to the particular /distribution/ of values within the initial UTxO
+--      set, the algorithm depletes all entries from the UTxO set /before/ it
+--      is able to pay for all requested outputs.
+--
+--      See: __'ErrUxtoFullyDepleted'__.
+--
+--  4.  The /number/ of UTxO entries needed to pay for the requested outputs
+--      would /exceed/ the upper limit specified by 'maximumInputCount'.
+--
+--      See: __'ErrMaximumInputCountExceeded'__.
+--
+-- = Motivating Principles
+--
+-- There are several motivating principles behind the design of the algorithm.
+--
+-- === Principle 1: Dust Management
+--
+-- The probability that random selection will choose dust entries from a UTxO
+-- set increases with the proportion of dust in the set.
+--
+-- Therefore, for a UTxO set with a large amount of dust, there's a high
+-- probability that a random subset will include a large amount of dust.
```

Comment:
the trade-off here, the tx will cost more as tx fee is bigger, but it has also benefits